File: test/test_trace_file.cpp
#define CATCH_CONFIG_MAIN // This tells Catch to provide a main() - only do this
                          // in one cpp file
#include <boost/filesystem.hpp>
#include <catch.hpp>
#include <thread>
#include <vector>
#include <trace_events.h>

#define private public
#include <trace_file.h>
#undef private

namespace bf = boost::filesystem;

TEST_CASE ("tracefile::create")
{
    {
        TraceFile tf ("./foo", TraceFileMode::WRITE);
    }
    bf::path p{ "./foo" };
    REQUIRE (bf::is_regular_file (p));
    REQUIRE (bf::remove (p));
}

TEST_CASE ("tracefile::metadata::rw")
{
    const char* p = "./foo";

    auto tid = std::this_thread::get_id ();

    EventVectorBuffer eb (42);

    TraceMetaData md_read;
    TraceMetaData md_write (eb, tid);
    {
        TraceFile tf (p, TraceFileMode::WRITE);
        tf.write_meta_data (md_write);
    }
    {
        TraceFile tf (p, TraceFileMode::READ);
        tf.read_meta_data (&md_read);
    }
    REQUIRE (md_write.thread_id () == md_read.thread_id ());
    REQUIRE (md_write.size () == md_read.size ());
    REQUIRE (bf::remove (p));
}

TEST_CASE("EventVectorBuffer")
{
    EventVectorBuffer buffer;
    AccessEvent ae1 (1, 0x1, 10, AccessType::STORE, MemoryLevel::MEM_LVL_L1);
    AccessEvent ae2 (2, 0x2, 20, AccessType::LOAD, MemoryLevel::MEM_LVL_L2);
    buffer.append(ae1);
    buffer.append(ae2);

    REQUIRE(buffer.size() == 2);

    auto iter = buffer.begin();

    REQUIRE(iter != buffer.end());
    REQUIRE (iter->time == ae1.time);
    REQUIRE (iter->address == ae1.address);
    REQUIRE (iter->ip == ae1.ip);
    REQUIRE (iter->access_type == ae1.access_type);
    REQUIRE (iter->memory_level == ae1.memory_level);

    iter++;
    REQUIRE (iter->time == ae2.time);
    REQUIRE (iter->address == ae2.address);
    REQUIRE (iter->ip == ae2.ip);
    REQUIRE (iter->access_type == ae2.access_type);
    REQUIRE (iter->memory_level == ae2.memory_level);

    iter++;
    REQUIRE (iter == buffer.end());

}

TEST_CASE("EventRingBuffer")
{
    EventRingBuffer buffer(2);
    AccessEvent ae1 (1, 0x1, 10, AccessType::STORE, MemoryLevel::MEM_LVL_L1);
    AccessEvent ae2 (2, 0x2, 20, AccessType::LOAD, MemoryLevel::MEM_LVL_L2);
    buffer.append(ae1);
    buffer.append(ae2);

    REQUIRE(buffer.size() == 2);

    auto iter = buffer.begin();

    REQUIRE(iter != buffer.end());
    REQUIRE (iter->time == ae1.time);
    REQUIRE (iter->address == ae1.address);
    REQUIRE (iter->ip == ae1.ip);
    REQUIRE (iter->access_type == ae1.access_type);
    REQUIRE (iter->memory_level == ae1.memory_level);

    iter++;
    REQUIRE (iter->time == ae2.time);
    REQUIRE (iter->address == ae2.address);
    REQUIRE (iter->ip == ae2.ip);
    REQUIRE (iter->access_type == ae2.access_type);
    REQUIRE (iter->memory_level == ae2.memory_level);

    iter++;
    REQUIRE (iter == buffer.end());

    buffer.append(ae2);
    iter = buffer.begin();

    REQUIRE (iter->time == ae2.time);
    REQUIRE (iter->address == ae2.address);
    REQUIRE (iter->ip == ae2.ip);
    REQUIRE (iter->access_type == ae2.access_type);
    REQUIRE (iter->memory_level == ae2.memory_level);
}

TEST_CASE ("tracefile::circular_buffer")
{
    const char* p = "./foobar";
    AccessEvent ae1 (1, 0x1, 10, AccessType::STORE, MemoryLevel::MEM_LVL_L1);
    AccessEvent ae2 (2, 0x2, 20, AccessType::LOAD, MemoryLevel::MEM_LVL_L2);
    EventRingBuffer eb (2);

    auto tid = std::this_thread::get_id ();
    eb.append (ae1);
    eb.append (ae2);

    REQUIRE (eb.size () == 2);

    {
        TraceFile tf (p, TraceFileMode::WRITE);
        tf.write (eb, TraceMetaData (eb, tid));
    }

    TraceFile tf (p, TraceFileMode::READ);
    auto result = tf.read<boost::circular_buffer<AccessEvent>> ();

    REQUIRE (result.size () == 2);

    REQUIRE (result[0].time == ae1.time);
    REQUIRE (result[0].address == ae1.address);
    REQUIRE (result[0].ip == ae1.ip);
    REQUIRE (result[0].access_type == ae1.access_type);
    REQUIRE (result[0].memory_level == ae1.memory_level);

    REQUIRE (result[1].time == ae2.time);
    REQUIRE (result[1].address == ae2.address);
    REQUIRE (result[1].ip == ae2.ip);
    REQUIRE (result[1].access_type == ae2.access_type);
    REQUIRE (result[1].memory_level == ae2.memory_level);

    REQUIRE (bf::remove (p));
}

TEST_CASE ("tracefile::vector")
{
    const char* p = "./foobaz";
    AccessEvent ae1 (1, 0x1, 10, AccessType::STORE, MemoryLevel::MEM_LVL_L1);
    AccessEvent ae2 (2, 0x2, 20, AccessType::LOAD, MemoryLevel::MEM_LVL_L2);
    EventVectorBuffer eb;

    auto tid = std::this_thread::get_id ();
    eb.append (ae1);
    eb.append (ae2);

    {
        TraceFile tf (p, TraceFileMode::WRITE);
        tf.write (eb, TraceMetaData (eb, tid));
    }

    TraceFile tf (p, TraceFileMode::READ);
    auto result = tf.read<std::vector<AccessEvent>> ();

    REQUIRE (result.size () == 2);

    REQUIRE (result[0].time == ae1.time);
    REQUIRE (result[0].address == ae1.address);
    REQUIRE (result[0].ip == ae1.ip);
    REQUIRE (result[0].access_type == ae1.access_type);
    REQUIRE (result[0].memory_level == ae1.memory_level);

    REQUIRE (result[1].time == ae2.time);
    REQUIRE (result[1].address == ae2.address);
    REQUIRE (result[1].ip == ae2.ip);
    REQUIRE (result[1].access_type == ae2.access_type);
    REQUIRE (result[1].memory_level == ae2.memory_level);

    REQUIRE (bf::remove (p));
}

TEST_CASE ("trace_events::memoryLevelFromPerf")
{
    union perf_mem_data_src data_src;
    data_src.mem_lvl = PERF_MEM_LVL_L1;
    data_src.mem_lvl_num = PERF_MEM_LVLNUM_L1;

    uint64_t t = data_src.val >> PERF_MEM_LVL_SHIFT;
    bool f = t & PERF_MEM_LVL_L1;
    REQUIRE (f);

    REQUIRE (memoryLevelFromPerf (data_src.mem_lvl) == MemoryLevel::MEM_LVL_L1);
}

File: test/test_trace_file.py
#! /usr/bin/env python3
import os.path
import unittest
import tracefile as tf

class TestAccessEvent(unittest.TestCase):
    def test_creation(self):
        a = tf.AccessEvent(1, 1, 42, tf.AccessType.LOAD, tf.MemoryLevel.MEM_LVL_L1)
        self.assertEqual(a.timestamp, 1)
        self.assertEqual(a.address, 1)
        self.assertEqual(a.ip, 42)
        self.assertEqual(a.type, tf.AccessType.LOAD)
        self.assertEqual(a.level, tf.MemoryLevel.MEM_LVL_L1)
        b = tf.AccessEvent()
        self.assertEqual(b.timestamp, 0)
        self.assertEqual(b.address, 0)
        self.assertEqual(b.ip, 0)
        self.assertEqual(b.type, tf.AccessType.NA)
        self.assertEqual(b.level, tf.MemoryLevel.MEM_LVL_NA)

class TestEventBuffer(unittest.TestCase):
    def test_add(self):
        buffer = tf.EventVectorBuffer()
        self.assertEqual(len(buffer), 0)
        buffer.append(tf.AccessEvent(1, 1, 1, tf.AccessType.LOAD, tf.MemoryLevel.MEM_LVL_HIT))
        self.assertEqual(len(buffer), 1)

    def test_range(self):
        buffer = tf.EventVectorBuffer()
        data = [
            tf.AccessEvent(1, 1, 1, tf.AccessType.LOAD, tf.MemoryLevel.MEM_LVL_HIT),
            tf.AccessEvent(2, 2, 2, tf.AccessType.STORE, tf.MemoryLevel.MEM_LVL_MISS),
            tf.AccessEvent(3, 3, 3, tf.AccessType.EXEC, tf.MemoryLevel.MEM_LVL_L1),
        ]

        for a in data:
            buffer.append(a)

        self.assertEqual(len(buffer), 3)

        for a, e in zip(buffer, data):
            self.assertEqual(a.timestamp, e.timestamp)
            self.assertEqual(a.address, e.address)
            self.assertEqual(a.ip, e.ip)
            self.assertEqual(a.type, e.type)
            self.assertEqual(a.level, e.level)

    def test_get_item(self):
        a1 = tf.AccessEvent(1, 1, 42, tf.AccessType.LOAD, tf.MemoryLevel.MEM_LVL_L1)
        a2 = tf.AccessEvent(2, 2, 44, tf.AccessType.STORE, tf.MemoryLevel.MEM_LVL_L2)
        buffer = tf.EventVectorBuffer()
        buffer.append(a1)
        buffer.append(a2)

        self.assertEqual(len(buffer), 2)
        self.assertEqual(a1.timestamp, buffer[0].timestamp)
        self.assertEqual(a1.address, buffer[0].address)
        self.assertEqual(a1.ip, buffer[0].ip)
        self.assertEqual(a1.type, buffer[0].type)
        self.assertEqual(a1.level, buffer[0].level)

        self.assertEqual(a2.timestamp, buffer[1].timestamp)
        self.assertEqual(a2.address, buffer[1].address)
        self.assertEqual(a2.ip, buffer[1].ip)
        self.assertEqual(a2.type, buffer[1].type)
        self.assertEqual(a2.level, buffer[1].level)

    def test_set_item(self):
        a1 = tf.AccessEvent(1, 1, 42, tf.AccessType.LOAD, tf.MemoryLevel.MEM_LVL_L1)
        a2 = tf.AccessEvent(2, 2, 44, tf.AccessType.STORE, tf.MemoryLevel.MEM_LVL_L2)
        buffer = tf.EventVectorBuffer()
        buffer.append(a1)
        buffer[0] = a2
        self.assertEqual(a2.timestamp, buffer[0].timestamp)
        self.assertEqual(a2.address, buffer[0].address)
        self.assertEqual(a2.ip, buffer[0].ip)
        self.assertEqual(a2.type, buffer[0].type)
        self.assertEqual(a2.level, buffer[0].level)

class TestTraceMetaData(unittest.TestCase):
    def test_creation(self):
        buffer = tf.EventVectorBuffer()
        md = tf.TraceMetaData(buffer, 1337)
        self.assertEqual(md.size(), 0)
        self.assertEqual(md.thread_id(), 1337)

class TestTraceFile(unittest.TestCase):
    def test_create(self):
        path = "./foo.txt"
        with tf.TraceFile(path, tf.TraceFileMode.WRITE) as file:
            self.assertEqual(path, file.path())

        self.assertTrue(os.path.isfile(path))
        info = os.stat(path)
        self.assertEqual(0, info.st_size)

    def test_write_read(self):
        path = "./foo.txt"
        a1 = tf.AccessEvent(1, 1, 42, tf.AccessType.LOAD, tf.MemoryLevel.MEM_LVL_L1)
        a2 = tf.AccessEvent(2, 2, 44, tf.AccessType.STORE, tf.MemoryLevel.MEM_LVL_L2)
        write_buffer = tf.EventVectorBuffer()
        write_buffer.append(a1)
        write_buffer.append(a2)
        md = tf.TraceMetaData(write_buffer, 100)
        with tf.TraceFile(path, tf.TraceFileMode.WRITE) as file:
            self.assertEqual(path, file.path())
            file.write(write_buffer, md)

        self.assertTrue(os.path.isfile(path))

        with tf.TraceFile(path, tf.TraceFileMode.READ) as file:
            self.assertEqual(path, file.path())
            read_buffer, read_md = file.read()

        self.assertEqual(md.size(), read_md.size())
        self.assertEqual(md.thread_id(), read_md.thread_id())

        for expect, current in zip(write_buffer, read_buffer):
            self.assertEqual(expect.timestamp, current.timestamp)
            self.assertEqual(expect.address, current.address)
            self.assertEqual(expect.ip, current.ip)
            self.assertEqual(expect.type, current.type)
            self.assertEqual(expect.level, current.level)


if __name__ == '__main__':
    unittest.main()